### DISCUSSION TITLE
Vector Tween Ranges

### DISCUSSION THREAD

**Discussion by konero** [opentoonz/opentoonz#6402] [Ideas]
### Is your feature request related to a problem? Please describe

Currently generating inbetween drawings from vector levels is tedious if adjustments to the key drawings are made later.

1. After generating inbetweens, user no longer knows which were the original key drawings
2. User must reselect the range of drawings
3. User must go through the interpolation dialog again

### Describe the solution you'd like

If OpenToonz had the ability to define a **tween range** similar to how Flash handles it we can store the interpolation type within the range and display to the user which drawings were the original key drawings. Since the interpolation type is stored we can generate the inbetween drawings with a command (or automatically) without hassle.

The user does the following to create a range:
1. Highlight some drawings
2. Initiate create tween range command
3. Select interpolation type

A range is created based on the first and last drawings in the selection, which are the key drawings, any drawings that were inbetween will become inbetween drawings of the range.

Take the following concept:

![image](https://user-images.githubusercontent.com/19820721/172380826-86640f7d-501e-411a-97d5-ade483905029.png)

Here we have three ranges, each range stores its own interpolation curve.

We need a way to:
1. Store the range
2. Visually display to the user where the ranges are

**Visual Concepts**
As perhaps it would be displayed to the user in the Level Strip.

![image](https://user-images.githubusercontent.com/19820721/172848772-ff498c6a-e062-4ca5-8a25-a3cf7ecbfb8b.png)

User can:
- Create Tween Range
- - See above description.
- Change Interpolation
- - Since the range is intended to store the interpolation curve, let the user change it later.
- Remove Tween Range
- - Only removes the range, does not delete drawings.

Assets:
- 📁 [curve_iconpack.zip](https://github.com/opentoonz/opentoonz/files/8870756/curve_iconpack.zip)

Related issues:
- #4310
- #1423

**Comment by beeheemooth** [comment]
I didn't quite understand: are you talking about key animation in Xsheet/Timeline or autoinbetweens in Level Strip here? Although in both cases, a way is really needed to mark the frames to which the tween has been applied (not to mention the automatic application of interpolation when editing these frames). I always forget where I applied what interpolation and insert notes everywhere, and this is very inconvenient.
From practice, I can say that the third option (Multiple non-linked tween ranges) is the most popular, as it allows you to make significant changes to the picture in the next frame, followed by the application of tween. This is what happens when you use the Plastic Tool with multiple skeletons.

**Comment by konero** [comment]
> I didn't quite understand: are you talking about key animation in Xsheet/Timeline or autoinbetweens in Level Strip here?

Apologies (I'll update the post soon), I'm talking about **auto inbetweens** generated in Level Strip, similar to Flash shape tweens.

> From practice, I can say that the third option (Multiple non-linked tween ranges) is the most popular, as it allows you to make significant changes to the picture in the next frame, followed by the application of tween. This is what happens when you use the Plastic Tool with multiple skeletons.

I see, thank you for the input! It's a difficult thing to conceptualize due to how Toonz works.

So, I think we can describe a range then as the non-linked type, which consists of two key drawings, any frames inbetween gets regenerated (manually, automatically), this would be the same as Flash shape tweens workflow I believe? And similar to CACANi.

![image](https://user-images.githubusercontent.com/19820721/172378512-79237318-142c-4862-a860-a68288d7050c.png)

As per the above image, there are three ranges, each would store its own interpolation curve when it was created.

So the problems we face are:
1. We need to store the range somehow
2. We need user interface elements to display the ranges to the user, in the X-sheet or Level Strip?

Does this seem reasonable?

**Comment by beeheemooth** [comment]
So, if you are talking about auto inbetweens, then it is logical that the display of frames with applied interpolation should occur in the Level Strip. But (again, from working practice), we know examples when different levels in the same time period can have the same (or different) set of interpolations, so that for orientation it is not necessary to constantly switch from one Level Strip to another, it is also logical that these frames were also displayed in the Xsheet. However, all these new elements will inevitably lead to changes in both the scene file and the vector level file. We discussed a similar feature request in the Tahoma discord. Manongjohn offered to write additional changes to a separate file in order to keep the Tahoma and Opentoonz files compatible.

Since we have a rather poor choice of interpolations for the Level Strip (there are only four of them), it would be possible to color the frames to which this or that interpolation was applied in four different colors.

**Comment by artisteacher** [comment]
I love the visual indicator in the level strip!

**Comment by ghost** [comment]
Personally, I also really like the ideas and icons to easily find your way around with auto-inbetweens, that would really be a big plus! (I also made a similar request as we can find in cacani here #3470, much less detailed). I would have a little reservation about the placement of arrows in the middle of the level-strip though. Since it's also used to display the drawings, I think it's a bit of an encroachment on the space and clarity on that side.

**Comment by RodneyBaker** [comment]
Converting this to Discussion so others can explore the provided information and consider the best way forward.
